express-api/src/services/keycloak/keycloakService.ts

```diff
@@ -0,0 +1,202 @@
+  // If not in PIMS, add it
+  // If PIMS has roles that aren't in Keycloak, remove them.
+};
+
+/**
+ * @description Fetch a list of groups from Keycloak and their associated role within PIMS
+ * @returns {IKeycloakRoles[]}  A list of roles from Keycloak.
+ */
+const getKeycloakRoles = async () => {
+  // Get roles available in Keycloak
+  const keycloakRoles: IKeycloakRolesResponse = await getRoles();
+  // Return the list of roles
+  return keycloakRoles.data;
+};
+
+/**
+ * @description Get information on a single Keycloak role from the role name.
+ * @param   {string}   roleName String name of role in Keycloak
+ * @returns {IKeycloakRole}  A single role object.
+ * @throws If the role does not exist in Keycloak.
+ */
+const getKeycloakRole = async (roleName: string) => {
+  // Get single role
+  const response: IKeycloakRole | IKeycloakErrorResponse = await getRole(roleName);
+  // Did the role exist? If not, it will be of type IKeycloakErrorResponse.
+  if (!keycloakRoleSchema.safeParse(response).success) {
+    const message = `keycloakService.getKeycloakRole: ${
+      (response as IKeycloakErrorResponse).message
+    }`;
+    logger.warn(message);
+    throw new Error(message);
+  }
+  // Return role info
+  return response;
+};
+
+/**
+ * @description Update a role that exists in Keycloak. Create it if it does not exist.
+ * @param   {string}   roleName String name of role in Keycloak
+ * @param   {string}   newRoleName The name to change the role name to.
+ * @returns {IKeycloakRole} The updated role information. Existing role info if cannot be updated.
+ * @throws  {Error} If the newRoleName already exists.
+ */
+const updateKeycloakRole = async (roleName: string, newRoleName: string) => {
+  const roleWithNameAlready: IKeycloakRole = await getRole(newRoleName);
+  // If it already exists, log the error and return existing role
+  if (keycloakRoleSchema.safeParse(roleWithNameAlready).success) {
+    const message = `keycloakService.updateKeycloakRole: Role ${newRoleName} already exists`;
+    logger.warn(message);
+    throw new Error(message);
+  }
+  const response: IKeycloakRole = await getRole(roleName);
+  // Did the role to be changed exist? If not, it will be of type IKeycloakErrorResponse.
+  let role: IKeycloakRole;
+  if (keycloakRoleSchema.safeParse(response).success) {
+    // Already existed. Update the role.
+    role = await updateRole(roleName, newRoleName);
+  } else {
+    // Didn't exist already. Add the role.
+    role = await createRole(newRoleName);
+  }
+
+  // Return role info
+  return role;
+};
+
+// TODO: Complete when user and role services are complete.
+const syncKeycloakUser = async () => {
+  // Does user exist in Keycloak?
+  // Get their existing roles.
+  // Does user exist in PIMS
+  // If user exists in PIMS...
+  // Update the roles in PIMS to match their Keycloak roles
+  // If they don't exist in PIMS...
+  // Add user and assign their roles
+};
+
+interface IKeycloakUsersFilter {
```

Comment:
Could just use the exported IDIRUserQuery from the keycloak package here instead

Reply:
Good point. I think I had that originally because of the package typing I didn't like. Replaced now.